src/blockListener.ts

```diff
@@ -1,11 +1,57 @@
-import networkClient from './networkClient';
-import { updateStats, verbose } from './utils';
-import { EthersObserverEvents } from './types';
-
-const blockListener = (): typeof networkClient.provider =>
-  networkClient.provider.on(EthersObserverEvents.Block, async (blockNumber) => {
-    verbose('Processing block #', blockNumber);
-    await updateStats({ latestBlock: blockNumber });
+import { output, getLastBlockNumber } from '~utils';
+import { Block, EthersObserverEvents } from '~types';
+import provider from '~provider';
+import { processNextBlock } from '~blockProcessor';
+
+/**
+ * Object storing blocks that have been either picked up by the block listener
+ * or missed blocks tracking
+ * The ability to index on block number made it much easier to ensure sequential
+ * processing of the blocks
+ */
+export const blocksMap: Record<number, Block | undefined> = {};
```

Comment:
1) I probably could and should be. But it seems to have no impact - the ingestor's been running for 42 hours straight and the memory usage is pretty much constant:
<img width="685" alt="image" src="https://github.com/JoinColony/block-ingestor/assets/112586815/a619a3f3-48a0-4bac-b617-2186c1fa8b51">

2) Good shout, amended 👍 

3) It's not a queue (although the logs suggest otherwise) since the order of the blocks is not guaranteed (for example, block 1020 can be picked up before block 1019, or in more complex scenario the order could be: 1020, 1019, 1021, 1018). Storing the blocks in an array would require a sorting round every time a new block is added. Instead, we have the map between seen block numbers and blocks data and a pointer variable that ensures the blocks are processed in the right order, one-by-one. I removed the comment entirely since it's not very helpful.